### PR TITLE
Fix login flow when session expires

### LINKED ISSUE #197: Login flow broken when sessions expire
### Versions

|             | Version |
| ----------- | ------- |
| Verdaccio   | 5.29.0 |
| This plugin | 6.0.8 |
| Node        | 20.10.0 |

### Environment

|                  | Name | Version |
| ---------------- | ---- | ------- |
| Package manager  | helm |         |
| Browser          | chrome |         |
| Operating system | Windows |         |

### Observed behavior

When a session expires in the web UI, the login flow doesn't work until I clear local storage. Specifically, after the login flow completes and I'm sent back to the Verdaccio landing page, it still says "Login" and I can't access anything. If I clear local storage manually and then go through the login flow, it works as expected.

### Expected behavior

Going through the login flow again after a session expires should work.

### Steps to reproduce

- Login on the web UI
- Wait an hour for the token to expire
- You should now see a "Login" button again as the UI detects your session is expired
- Click "Login" and go through the auth flow
- When you return to the landing page, you'll still see the "Login" button and are unable to access content

### Additional context

I did some investigating and found a few things that might help:
1. When your session expires, the `token` value remains in local storage. The only code path I found that removes this value is [clicking the "Logout" button](https://github.com/n4bb12/verdaccio-github-oauth-ui/blob/06cc80547c9f2c19f119312d1e4ae84a7641acf2/src/client/plugin/init.ts#L55), but unfortunately this button isn't visible on the page when your session has expired.
2. Upon completing the login flow, [the `saveCredentials` call](https://github.com/n4bb12/verdaccio-github-oauth-ui/blob/06cc80547c9f2c19f119312d1e4ae84a7641acf2/src/client/plugin/init.ts#L21) that would save the new values into local storage does not get invoked. This is because the `isLoggedIn` function returns true a few lines before.
3. As mentioned in the previous line, `isLoggedIn` returns true when a session has expired. It doesn't seem to check the expiry date in the JWT token, [it simply checks that the token exists](https://github.com/n4bb12/verdaccio-github-oauth-ui/blob/06cc80547c9f2c19f119312d1e4ae84a7641acf2/src/client/plugin/credentials.ts#L25).

### PR DESCRIPTION
**Pull request by thecodeboss**
## Summary
Clear local storage if we detect that the token is expired. Fixes #197.

## Notes
* The `isTokenExpired` treats a token as expired if we're within 30 seconds of expiry. This matches [the logic Verdaccio itself uses](https://github.com/verdaccio/verdaccio/blob/007a7bd76c0391d266fc89e5d5a5fa7192d93be2/packages/ui-components/src/utils/token.ts#L5), which is important since the UI does not match up otherwise.
* I used the `parseJwt` method [from this Stack Overflow answer](https://stackoverflow.com/a/38552302/1935971). I noticed the client code in this repo did not depend on any libraries, so I kept it that way.

### CODE REVIEW

File: src/client/plugin/lib.ts [hunk 1]
```diff
@@ -15,6 +15,19 @@ export function parseCookies(cookieStr: string) {
...
 
+export function parseJwt(token: string) {
+  const base64Url = token.split(".")[1]
+  const base64 = base64Url.replace(/-/g, "+").replace(/_/g, "/")
+  const jsonPayload = decodeURIComponent(
+    atob(base64)
+      .split("")
+      .map((c) => "%" + ("00" + c.charCodeAt(0).toString(16)).slice(-2))
+      .join(""),
+  )
```
Comment (n4bb12) [inline src/client/plugin/lib.ts:26]: There is a lot going on here that isn't clear to me. Could you add a comment or link?

**Review by n4bb12 — COMMENTED** [review]
Thank you very much.

File: src/client/plugin/lib.ts (hunk 1 above)
Comment (n4bb12) [reply src/client/plugin/lib.ts]: Thank you. Could you add a unit test for this function?

Comment (thecodeboss) [reply src/client/plugin/lib.ts]: Hey @n4bb12, apologies for the delay, I was on vacation for the past week. I just added a unit test if you could take another look :)

**Comment by n4bb12** [comment]
Great, thank you

**Review by n4bb12 — APPROVED** [review]

**Pull request merged**